### PR TITLE
WIP: Read schema defaults when generating field using EntityContext.

### PR DESCRIPTION
**Pull request by ADmad**
In 2.x when creating an input if field value was not set in request data the default value available from table schema was set as field value (if an explicit default value was not provided).

This patch implements the same feature.

I'll add more tests if the change is approved.

### CODE REVIEW

File: src/View/Form/EntityContext.php [hunk 1]
```diff
@@ -201,16 +201,26 @@ public function isCreate()
...
      * @param string $field The dot separated path to the value.
+     * @param array $options Options:
+     *   - `default`: Default value to return if no value found in request
+     *     data or entity.
+     *   - `schemaDefault`: Boolen indicating whether default value from table
+     *     schema should be used if it's not explicitly provided.
      * @return mixed The value of the field or null on a miss.
      */
-    public function val($field)
+    public function val($field, $options = [])
```
Comment (dereuromark) [inline src/View/Form/EntityContext.php]: Should we array typehint it?

Comment (ADmad) [inline src/View/Form/EntityContext.php]: Typehints have caused us quite a bit of grief, so for methods like this which are generally not going to be directly used by users i would prefer not adding typhints.

**Comment by ADmad** [comment]
The jobs failing for sqlite are due to a bug in `SqliteSchema`. For table columns created with `DEFAULT NULL` when the schema is introspected it sets `'default'` option as string `'NULL'` instead of `null`.

@chinpei215 is gonna help fix that along with other changes he's working on for schema classes :)

File: src/View/Form/EntityContext.php [hunk 2]
```diff
@@ -230,6 +248,26 @@ public function val($field)
...
+     * @param \Cake\Datasource\EntityInterface $entity The entity.
+     * @return mixed
+     */
+    protected function _schemaDefault($field, $entity)
+    {
+        $table = $this->_getTable($entity);
+        if ($table === false) {
+            return null;
+        }
+        $defaults = $table->schema()->defaultValues();
```
Comment (markstory) [inline src/View/Form/EntityContext.php]: Should we add a method to get the default value for a specific column? `schema()->defaultValue($field)`?

File: src/View/Helper/FormHelper.php [hunk 3]
```diff
@@ -2407,7 +2407,11 @@ protected function _initInputField($field, $options = [])
             unset($options['value']);
         }
         if (!isset($options['val'])) {
-            $options['val'] = $context->val($field);
+            $valOptions = [
+                'default' => isset($options['default']) ? $options['default'] : null,
+                'schemaDefault' => isset($options['schemaDefault']) ? $options['schemaDefault'] : true,
+            ];
+            $options['val'] = $context->val($field, $valOptions);
```
Comment (markstory) [inline src/View/Helper/FormHelper.php]: Could we get some integration tests for the form helper. This new behavior has no tests.

File: src/View/Form/EntityContext.php (hunk 2 above)
Comment (ADmad) [inline src/View/Form/EntityContext.php]: We should.

File: src/View/Helper/FormHelper.php (hunk 3 above)
Comment (ADmad) [inline src/View/Helper/FormHelper.php]: Yup, I will. Was waiting for green signal on the idea in general before putting in more effort :)

Comment (markstory) [inline src/View/Helper/FormHelper.php]: I think the idea is reasonable. There is even an open issue for this #5065

**Comment by markstory** [comment]
This should probably go into `3.next` as to not surprise people who upgrade to a 'bugfix' release.

**Comment by ADmad** [comment]
@markstory I'll retarget the PR

File: src/View/Helper/FormHelper.php (hunk 3 above)
Comment (ADmad) [inline src/View/Helper/FormHelper.php]: I did remember an issue being submitted for this but couldn't find it :)

**Pull request closed without merging**